### PR TITLE
V2.2.0 alpha

### PR DESCRIPTION
**Pull request by dragonmantank**
This contains all the work discussed for what was originally v3, but in a backwards-compatible release. Existing test suite was used and almost entirely untouched to make sure that existing users could update to this minor release without breaking anything. Only test changes involved some text formatting (which was corrected for PSR-2) and some badly formed tests (which were made more explicit in what they test).

v3.0.0 will remove all the deprecated features, and only modify a few end-user methods to be more clear that were not easily to deprecate in this release. That will come soon after this branch looks good.

Broad overview:
- All the underlying API interactions now moved to `Nexmo\Client\APIResource`
  - Provides more centralized error handling
  - Allows easier access to raw requests and responses instead of shuffling around on entities
- Entities are closer to being Value Objects, and working toward immutability
- Vocabulary is more standardized, with REST-ish method names deprecated
- Voice and SMS are completely revamped, with helpers for building messages and NCCOs

### CODE REVIEW

**Comment by lornajane** [comment]
We have a bunch of reviewers tagged here (hurrah for a talented team). It's a huge change and I'm not really sure where to start. What would you like us to do here to help move this forward?

**Comment by dragonmantank** [comment]
> We have a bunch of reviewers tagged here (hurrah for a talented team). It's a huge change and I'm not really sure where to start. What would you like us to do here to help move this forward?

For the most part in this branch it is if the depredations look and make sense. Some of this, like hydration and some new functions are being made available as an easier upgrade path to what will be v3, but I want to especially make sure it doesn't look like we're deprecating anything incorrectly (like removing support for a data type or structure that we don't have a proper replacement for).

File: composer.json [hunk 1]
```diff
@@ -34,7 +34,8 @@
     "php-http/client-implementation": "^1.0",
     "zendframework/zend-diactoros": "^1.8.4 || ^2.0",
     "lcobucci/jwt": "^3.2",
-    "ocramius/package-versions": "^1.4"
+    "ocramius/package-versions": "^1.4",
+    "phpstan/phpstan": "^0.12.11"
```
[the remark this answers is not in the document]
Comment (dragonmantank) [reply composer.json:38]: It should be, I'll move that

**Comment by lornajane** [comment]
I tried some of the existing code samples and although the library had more array-oriented and object-method-centric ways to do things, all of them now give me a lot of deprecation notices. What does the "recommended" way to use this library look like now?

**Comment by dragonmantank** [comment]
> I tried some of the existing code samples and although the library had more array-oriented and object-method-centric ways to do things, all of them now give me a lot of deprecation notices. What does the "recommended" way to use this library look like now?

Great. All the snippets should work, but many will generate the deprecation notices as a large portion of the snippets rely on raw arrays. The intent was to not stop the code, but let the user know this stuff was going to be going away/changing.

Out of the notices, did they seem clear enough on what one would need to do in the cases to "upgrade" them to get rid of the deprecation notice??

**Comment by lornajane** [comment]
Not really, I pulled a code example from the README, and it says I can't access as an array or set a client directly, but I am not sure how to find out what I should have done instead.

**Comment by dragonmantank** [comment]
Which example, just so I can test?

Array access and looping over the main client is going away, so I want to make sure it's communicated properly for existing customers who are upgrading.

**Comment by lornajane** [comment]
"Starting a Verification", there are two ways to do it, both generate deprecation output but I wasn't sure what I should have done instead.

**Comment by dragonmantank** [comment]
That is perfect. Verification did what it was supposed to (complained, but ultimately worked), just generated a lot of ancillary deprecation notices over things the _user_ doesn't normally do, because internally it's using the old way.

What I'll need to do is go through and refactor to cut down on the non-user-generated notices. Thanks!

**Pull request closed without merging**